### ISSUE TITLE
I use 'Spring Cloud Greenwich.RELEASE' and my api declare @ReactiveFeignClient, but micro-service can not get eureka server list for service

### ISSUE THREAD

**Issue by haycco** [Playtika/feign-reactive#38]
Hi,
I using  Spring Cloud the version is **Greenwich.RELEASE** and feign-reactive version is '1.0.15', my service api declare using   **@ReactiveFeignClient** and use **@Autowired** inject to api client instance, when my eureka client micro-service invoke the api remote method, it can not got eureka registry services list. when I  debug found some problem need to help..

package com.netflix.loadbalancer;
LoadBalancerContext
```
ILoadBalancer lb = this.getLoadBalancer();
        if (host == null) {
            if (lb != null) {
                Server svc = lb.chooseServer(loadBalancerKey);
                if (svc == null) {
                    throw new ClientException(ErrorType.GENERAL, "Load balancer does not have available server for client: " + this.clientName);
                }

                host = svc.getHost();
                if (host == null) {
                    throw new ClientException(ErrorType.GENERAL, "Invalid Server for :" + svc);
                }

                logger.debug("{} using LB returned Server: {} for request {}", new Object[]{this.clientName, svc, original});
                return svc;
            }
```
the **svc** is null!

until to manual configure bootstap.yaml like this:
```
xxx-service:
  ribbon:
    listOfServers: 'http://127.0.0.1:9002'
```
that **svc** can got not null.
but spring cloud micro-services too many and using docker, so that can not manual configure client, i think it is auto fetch eureka registry service.

under is my eureka client config,
```
eureka:
  instance:
    prefer-ip-address: true
    metadata-map:
      user.name: ${spring.security.user.name}
      user.password: ${spring.security.user.password}
  client:
    registerWithEureka: true
    fetchRegistry: true
    registryFetchIntervalSeconds: 5
    service-url:
      defaultZone: http://${spring.security.user.name}:${spring.security.user.password}@localhost:8761/eureka/
```
WebClientConfig.java
```
import com.netflix.client.ClientFactory;
import com.netflix.client.DefaultLoadBalancerRetryHandler;
import com.netflix.client.RequestSpecificRetryHandler;
import com.netflix.hystrix.HystrixCommandGroupKey;
import com.netflix.hystrix.HystrixCommandKey;
import com.netflix.hystrix.HystrixCommandProperties;
import com.netflix.hystrix.HystrixObservableCommand;
import com.netflix.loadbalancer.reactive.LoadBalancerCommand;
import feign.Response;
import feign.codec.ErrorDecoder;
import org.springframework.cloud.client.loadbalancer.LoadBalanced;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.http.client.reactive.ReactorClientHttpConnector;
import org.springframework.web.reactive.function.client.WebClient;
import reactivefeign.ReactiveOptions;
import reactivefeign.cloud.CloudReactiveFeign;
import reactivefeign.cloud.LoadBalancerCommandFactory;
import reactivefeign.retry.BasicReactiveRetryPolicy;
import reactivefeign.retry.ReactiveRetryPolicy;
import reactivefeign.webclient.WebReactiveOptions;

@Configuration
public class WebClientConfig {

    @Bean
    public LoadBalancerCommandFactory balancerCommandFactory(){
        return serviceName ->
                LoadBalancerCommand.builder()
                        .withLoadBalancer(ClientFactory.getNamedLoadBalancer(serviceName))
                        .withRetryHandler(new RequestSpecificRetryHandler(true, true,
                                new DefaultLoadBalancerRetryHandler(1, 0, true), null))
                        .build();
    }

    @Bean
    CloudReactiveFeign.SetterFactory setterFactory() {
        return (target, methodMetadata) -> {
            String groupKey = target.name();
            HystrixCommandKey commandKey = HystrixCommandKey.Factory.asKey(methodMetadata.configKey());
            return HystrixObservableCommand.Setter
                    .withGroupKey(HystrixCommandGroupKey.Factory.asKey(groupKey))
                    .andCommandKey(commandKey)
                    .andCommandPropertiesDefaults(HystrixCommandProperties.Setter()
                            .withExecutionTimeoutEnabled(false)
                    );
        };
    }

    @Bean
    @LoadBalanced
    public WebClient.Builder loadBalancedWebClientBuilder() {
        return WebClient.builder().clientConnector(new ReactorClientHttpConnector());
    }

    @Bean
    public ReactiveOptions reactiveOptions(){
        return new WebReactiveOptions.Builder().setReadTimeoutMillis(500).build();
    }

    @Bean
    public ReactiveRetryPolicy reactiveRetryPolicy(){
        return BasicReactiveRetryPolicy.retryWithBackoff(1, 10);
    }

    @Bean
    public feign.codec.ErrorDecoder reactiveStatusHandler(){
        return new ErrorDecoder() {
            @Override
            public Exception decode(String s, Response response) {
                return null;
            }
        };
    }

}
```

when I remove
```
xxx-service:
  ribbon:
    listOfServers: 'http://127.0.0.1:9002'
```
the problem is throw exception:
```
Caused by: com.netflix.client.ClientException: Load balancer does not have available server for client: default
	at com.netflix.loadbalancer.LoadBalancerContext.getServerFromLoadBalancer(LoadBalancerContext.java:483) ~[ribbon-loadbalancer-2.3.0.jar:2.3.0]
	at com.netflix.loadbalancer.reactive.LoadBalancerCommand$1.call(LoadBalancerCommand.java:184) [ribbon-loadbalancer-2.3.0.jar:2.3.0]
	... 168 common frames omitted
```
this config param not work
```
ribbon:
  eureka:
    enabled: false/true
```

if loadbalance what should I config,,,,so sad!
Thanks.

**Comment by kptfh** [comment]
Reactive Feign already has integration with Spring Cloud services (Ribbon, Hystrix).
You just need to add [cloud module](https://github.com/Playtika/feign-reactive/tree/develop/feign-reactor-cloud) to classpath and Feign Client will be auto configured as Cloud ready.
Check this example with [cloud autoconfiguration ](https://github.com/Playtika/feign-reactive/tree/develop/feign-reactor-cloud)
or use this class for [manual configuration](https://github.com/Playtika/feign-reactive/blob/develop/feign-reactor-cloud/src/main/java/reactivefeign/cloud/CloudReactiveFeign.java)

**Comment by haycco** [comment]
@kptfh 
manual configuration code
```
DefaultClientConfigImpl clientConfig = new DefaultClientConfigImpl();
        clientConfig.loadDefaultValues();
        clientConfig.setProperty(CommonClientConfigKey.NFLoadBalancerClassName, BaseLoadBalancer.class.getName());
        ILoadBalancer lb = ClientFactory.registerNamedLoadBalancerFromclientConfig("xxx-service", clientConfig);
        lb.addServers(asList(new Server("localhost", 9002)));
```
is work ok.
But...how to auto inject servers list value?

I already add **cloud module** to classpath, and inject to Reactive Feign Client, but remote invoke is wrong, if not manual configuration. 
Do you know how to auto inject eureka registry client server list?

I see the test case is manual configuration, but not auto. could you show case code?
Thanks!

**Comment by haycco** [comment]
@kptfh 
I have solve it autoconfiguration..when I override the bean like under code example:
```
@Bean
    public LoadBalancerCommandFactory balancerCommandFactory() {
        return serviceName -> {
            IRule rule = new AvailabilityFilteringRule();
            ServerList<DiscoveryEnabledServer> list = new DiscoveryEnabledNIWSServerList(serviceName);
            ServerListFilter<DiscoveryEnabledServer> filter = new ZoneAffinityServerListFilter<DiscoveryEnabledServer>();
            ZoneAwareLoadBalancer<DiscoveryEnabledServer> lb = LoadBalancerBuilder.<DiscoveryEnabledServer>newBuilder()
                    .withDynamicServerList(list)
                    .withRule(rule)
                    .withServerListFilter(filter)
                    .buildDynamicServerListLoadBalancer();

            System.out.println(lb);
            DiscoveryEnabledServer server = (DiscoveryEnabledServer) lb.chooseServer();

        return LoadBalancerCommand.builder()
                .withLoadBalancer(ClientFactory.getNamedLoadBalancer(serviceName))
                .withRetryHandler(new RequestSpecificRetryHandler(true, true,
                        new DefaultLoadBalancerRetryHandler(1, 0, true), null))
                .withServer(server)
                .build();
         };
    }
```
Every api client can scan and registar...
and now is work fine!

But some strange problem....
```
package com.netflix.loadbalancer;

DynamicServerListLoadBalancer

@VisibleForTesting
    public void updateListOfServers() {
        List<T> servers = new ArrayList<T>();
        if (serverListImpl != null) {
            servers = serverListImpl.getUpdatedListOfServers();
            LOGGER.debug("List of Servers for {} obtained from Discovery client: {}",
                    getIdentifier(), servers);

            if (filter != null) {
                servers = filter.getFilteredListOfServers(servers);
                LOGGER.debug("Filtered List of Servers for {} obtained from Discovery client: {}",
                        getIdentifier(), servers);
            }
        }
        updateAllServerList(servers);
    }

```
the above code can not refresh servers list form discovery client.
it is a bug?

**Comment by kptfh** [comment]
Thanks for your efforts. Just fixed bug in load balancer and now making 1.1.16 release. Here is the fully functioning sample project with eureka, web-flux and reactive feign [sample projects](https://github.com/kptfh/feign-reactive-sample). Now it's using 1.1.16-SNAPSHOT. After release will update it to 1.1.16.
